### DISCUSSION TITLE
CoAPS observe

### DISCUSSION THREAD

**Discussion by mkonnov** [openthread/openthread#9861] [Q&A]
Hey everyone,

I need your assistance in understanding the possibiliry of observing the resources over CoAPS.
From what I see in CLI of a client, there's no such option (`OT_COAP=ON` `OT_COAP_OBSERVE=ON` `OT_COAPS=ON`).
So for `coap` we see the following command set
```
> coap

cancel
delete
get
observe
parameters
post
put
resource
set
start
stop
```
But `coaps` lacks of `observe` and `cancel` commands
```
> coaps

connect
delete
disconnect
get
post
put
resource
set
start
stop
x509
```

Ok, I started to look into the CoAPS implementation and it seems that `CoapSecure` class just inherits the `CoapBase` that has the `SendMessage` method which in theory should handle the transmission over DTLS socket.

Q1. Is it possible at all to use CoAPS observe using OT SDK ?
Q2. If so - is there any code reference I may look at ?
Q2. Is it anyhow possible to use the OT CLI as a client for CoAPS observe ? Likely not but maybe I'm missing smth.

**Comment by abtink** [comment]
> Q1. Is it possible at all to use CoAPS observe using OT SDK ?

Technically yes. As you mentioned, `CoapBase` provides this functionality and is used by both `Coap` and `CoapSecure` classes. For app layer use, the observer APIs needs to be implemented/exposed as new public OT APIS under `otCoapSecure`  (and as CLI commands if needed).

One point, the observe functionality was contributed by intrested folks (from https://github.com/openthread/openthread/pull/4396). It is not a functionaility required/used by Thread stack itself, so it may not be as well-tested and validated as other features. 

> Q2. Is it anyhow possible to use the OT CLI as a client for CoAPS observe ? Likely not but maybe I'm missing smth.

CLI `coap` provides `observe` command. It is not supported by `coaps` but should be doable. Please feel free to contribute this if you want.  

In general CLI `coap` command are intended for quick testing. Public APIs are my recommendation for next layer/application code to use.

**Comment by mkonnov** [comment]
Thanks for your quick reply @abtink .
One question arose during the discovery of the topic though.

Do I need to change the code of openthread to acheive the successful CoAPS observe ?

Let's consider 2 parties in CoAPS communication: server and client.
For client part it's pretty obvious that we have to construct the message accordingly (by using `otCoapMessageAppendObserveOption`) and send to the server it's connected to.

But is it enough to use the vanilla openthread from mainline in order to server respond accordingly and the client receive the responses every time the server updates the resource ? In other words - is there any code changes in openthead needed for this ?

Thanks.

**Reply by abtink** [reply]
> But is it enough to use the vanilla openthread from mainline in order to server respond accordingly and the client receive the responses every time the server updates the resource ? In other words - is there any code changes in openthead needed for this ?


I expect it would work. But I am not familiar with the existing "observer" CoAP functionality to know for sure..

Suggest maybe trying it out, or read the code (maybe the PR I mentioned above that added support for this feature originally) to see what behaviors are already implemented and if anything is missing.

**Reply by mkonnov** [reply]
Ok, I'll try out and make you know.
Just asking not to spend a working day on a something that don't have to work.

The PR you pointed at is the PR that adds the observe functionality to the unsecure CoAP but my goal is to make the CoAPS observe functionality work.

Anyway, thanks for the directions.
Will get back with the results.
